### PR TITLE
Change allocator gRPC response state to gRPC error status

### PR DESCRIPTION
**Pull request by pooneh-m**
**What type of PR is this?**
/kind breaking

**What this PR does / Why we need it**:
Allocator service gRPC API is changed to instead of returning response.status.state as allocated, unallocated or conflict return gRPC error status in case of failure and drop response.status.state API from the response.

**Which issue(s) this PR fixes**:
Closes https://github.com/googleforgames/agones/issues/1040

### CODE REVIEW

File: cmd/allocator/main.go [hunk 1]
```diff
@@ -277,8 +277,8 @@ func (h *serviceHandler) Allocate(ctx context.Context, in *pb.AllocationRequest)
 		logger.Errorf("internal server error - Bad GSA format %v", resultObj)
 		return nil, status.Errorf(codes.Internal, "internal server error- Bad GSA format %v", resultObj)
 	}
-	response := converters.ConvertGSAV1ToAllocationResponseV1Alpha1(allocatedGsa)
-	logger.WithField("response", response).Infof("allocation response is being sent")
+	response, err := converters.ConvertGSAV1ToAllocationResponseV1Alpha1(allocatedGsa)
+	logger.WithField("response", response).WithField("err", err).Infof("allocation response is being sent")
```
Comment (aLekSer) [inline cmd/allocator/main.go:281]: Could we use https://godoc.org/gopkg.in/Sirupsen/logrus.v0#WithError here?
```suggestion
	logger.WithField("response", response).WithError(err).Infof("allocation response is being sent")
```

File: pkg/allocation/converters/converter.go [hunk 2]
```diff
@@ -206,28 +210,14 @@ func convertAllocationPortsV1Alpha1ToAgonesPortsV1(in []*pb.AllocationResponse_G
...
 	switch in {
 	case allocationv1.GameServerAllocationAllocated:
-		return pb.AllocationResponse_Allocated
+		return nil
 	case allocationv1.GameServerAllocationUnAllocated:
-		return pb.AllocationResponse_UnAllocated
+		return status.Error(codes.ResourceExhausted, "there isn't any available Game Server to allocate")
 	case allocationv1.GameServerAllocationContention:
-		return pb.AllocationResponse_Contention
+		return status.Error(codes.Aborted, "too many concurrent request has overwhelmed the system")
```
Comment (aLekSer) [inline pkg/allocation/converters/converter.go:220]:
```suggestion
		return status.Error(codes.Aborted, "too many concurrent requests has overwhelmed the system")
```

File: pkg/allocation/converters/converter.go [hunk 3]
```diff
@@ -206,28 +210,14 @@ func convertAllocationPortsV1Alpha1ToAgonesPortsV1(in []*pb.AllocationResponse_G
...
 // convertStateV1ToAllocationStateV1Alpha1 converts GameServerAllocationState V1 (GSA) to AllocationResponse_GameServerAllocationState
-func convertStateV1ToAllocationStateV1Alpha1(in allocationv1.GameServerAllocationState) pb.AllocationResponse_GameServerAllocationState {
+func convertStateV1ToError(in allocationv1.GameServerAllocationState) error {
 	switch in {
 	case allocationv1.GameServerAllocationAllocated:
-		return pb.AllocationResponse_Allocated
+		return nil
 	case allocationv1.GameServerAllocationUnAllocated:
-		return pb.AllocationResponse_UnAllocated
+		return status.Error(codes.ResourceExhausted, "there isn't any available Game Server to allocate")
```
Comment (aLekSer) [inline pkg/allocation/converters/converter.go:218]:
```suggestion
		return status.Error(codes.ResourceExhausted, "there is no available Game Server to allocate")
```
or could be 
`"there aren't any available Game Servers to allocate"`

File: pkg/allocation/converters/converter.go [hunk 4]
```diff
@@ -206,28 +210,14 @@ func convertAllocationPortsV1Alpha1ToAgonesPortsV1(in []*pb.AllocationResponse_G
...
-	switch in {
-	case pb.AllocationResponse_Allocated:
-		return allocationv1.GameServerAllocationAllocated, true
-	case pb.AllocationResponse_UnAllocated:
-		return allocationv1.GameServerAllocationUnAllocated, true
-	case pb.AllocationResponse_Contention:
-		return allocationv1.GameServerAllocationContention, true
-	}
-	return allocationv1.GameServerAllocationUnAllocated, false
+	return status.Error(codes.Unknown, "Unknown issue")
```
Comment (aLekSer) [inline pkg/allocation/converters/converter.go:222]: I think here should be also lower case to be consistent with above
```suggestion
	return status.Error(codes.Unknown, "unknown issue")
```

File: test/e2e/allocator_test.go [hunk 5]
```diff
@@ -393,3 +389,11 @@ func generateTLSCertPair(t *testing.T, host string) ([]byte, []byte) {
 
 	return pemPubBytes, pemPrivBytes
 }
+
+func validateAllocatorResponse(t *testing.T, resp *pb.AllocationResponse) {
+	t.Helper()
+	assert.Greater(t, len(resp.Ports), 0)
```
Comment (aLekSer) [inline test/e2e/allocator_test.go:395]: Could here be assert.Equal(..., 1) or other specific value ?

File: test/e2e/allocator_test.go [hunk 6]
```diff
@@ -393,3 +389,11 @@ func generateTLSCertPair(t *testing.T, host string) ([]byte, []byte) {
 
 	return pemPubBytes, pemPrivBytes
 }
+
+func validateAllocatorResponse(t *testing.T, resp *pb.AllocationResponse) {
+	t.Helper()
```
Comment (aLekSer) [inline test/e2e/allocator_test.go:394]: Something like below could be added:
```suggestion
	t.Helper()
	assert.NotNil(t, resp, "response should be present")
```

File: test/e2e/allocator_test.go [hunk 7]
```diff
@@ -183,13 +181,11 @@ func TestAllocatorCrossNamespace(t *testing.T) {
 			logrus.WithError(err).Info("failing Allocate request")
```
Comment (aLekSer) [inline test/e2e/allocator_test.go:181]: Line 173 above 
```
   logrus.WithError(err).Info("failing http request")
```
Should be `"failing grpc.Dial"` as in other places.

File: test/e2e/allocator_test.go (hunk 5 above)
Comment (pooneh-m) [reply test/e2e/allocator_test.go:395]: For this test >0 is enough for validation.

File: test/e2e/allocator_test.go (hunk 7 above)
Comment (pooneh-m) [reply test/e2e/allocator_test.go:181]: No this is actually failing the allocate request. grpc.Dial is successful at this point.

File: cmd/allocator/main.go (hunk 1 above)
Comment (pooneh-m) [reply cmd/allocator/main.go]: Thanks. Done.

File: pkg/allocation/converters/converter.go (hunk 2 above)
Comment (pooneh-m) [reply pkg/allocation/converters/converter.go]: Done.

File: pkg/allocation/converters/converter.go (hunk 3 above)
Comment (pooneh-m) [reply pkg/allocation/converters/converter.go]: Done.

File: pkg/allocation/converters/converter.go (hunk 4 above)
Comment (pooneh-m) [reply pkg/allocation/converters/converter.go]: Done

File: test/e2e/allocator_test.go (hunk 6 above)
Comment (pooneh-m) [reply test/e2e/allocator_test.go:394]: Done

File: test/e2e/allocator_test.go (hunk 7 above)
Comment (aLekSer) [reply test/e2e/allocator_test.go:181]: I mean not in this exact place but above:
https://github.com/googleforgames/agones/blob/6a02b31331501ae5ed9bf844107af47ca254e8c4/test/e2e/allocator_test.go#L175

File: test/e2e/allocator_test.go (hunk 5 above)
Comment (aLekSer) [reply test/e2e/allocator_test.go:398]: OK, got it

File: go.sum [hunk 8]
```diff
@@ -290,6 +290,7 @@ google.golang.org/grpc v1.17.0/go.mod h1:6QZJwpn2B+Zp71q/5VxRsJ6NXXVCE5NRUHRo+f3
 google.golang.org/grpc v1.19.0/go.mod h1:mqu4LbDTu4XGKhr4mRzUsmM4RtVoemTSY81AxZiDr8c=
 google.golang.org/grpc v1.20.1 h1:Hz2g2wirWK7H0qIIhGIqRGTuMwTE8HEKFnDZZ7lm9NU=
 google.golang.org/grpc v1.20.1/go.mod h1:10oTOabMzJvdu6/UiuZezV6QK5dSlG84ov/aaiqXj38=
+google.golang.org/grpc v1.29.1 h1:EC2SB8S04d2r73uptxphDSUG+kTKVgjRPF+N3xpxRB4=
```
Comment (aLekSer) [inline go.sum:293]: I am curious why `go.mod` was not changed?
It would still contain as in master?
```
	google.golang.org/grpc v1.20.1
```

Comment (pooneh-m) [reply go.sum:293]: Hmm. I am not sure. I just did go mod tidy and it went away.

File: test/e2e/allocator_test.go (hunk 7 above)
Comment (pooneh-m) [reply test/e2e/allocator_test.go:181]: Done.

File: go.sum (hunk 8 above)
Comment (pooneh-m) [reply go.sum]: Done.

Comment (aLekSer) [reply go.sum]: Thanks, now it is better.

**Comment by pooneh-m** [comment]
> Approving, PR looks good, no more requests from my side.

Thanks for your thorough review.

**Pull request merged**